### PR TITLE
Build selection UI

### PR DESCRIPTION
**Pull request by adammw**
Screen Shot 2019-09-17 at 6 41 23 pm

Work in progress.

### References
- Jira link: 

### Risks
- Low/Med/High: thing that could happen

### CODE REVIEW

File: app/controllers/builds_controller.rb [hunk 1]
```diff
@@ -10,6 +10,13 @@ class BuildsController < ApplicationController
 
   def index
     @builds = scope.order('id desc')
+
+    if params[:reference].present?
+      git_sha = current_project.repository.commit_from_ref(params[:reference])
+      raise "Commit not found for reference: #{params[:reference]}" unless git_sha.present?
```
Comment (grosser) [inline app/controllers/builds_controller.rb]: unknown ref should return empty results ?

File: app/helpers/builds_helper.rb [hunk 2]
```diff
@@ -21,14 +21,6 @@ def git_ref_and_sha_for(build, make_link: false)
...
-    if job = build.docker_build_job
-      status_badge(job.status)
-    elsif build.docker_repo_digest
-      status_badge("succeeded")
-    elsif (Job::VALID_STATUSES - ["succeeded"]).include?(build.external_status)
-      status_badge(build.external_status)
-    else
-      "not built"
-    end
+    status_badge build.status
```
Comment (grosser) [inline app/helpers/builds_helper.rb]: what will no status look like ?

File: app/models/build.rb [hunk 3]
```diff
@@ -17,13 +17,18 @@ class Build < ActiveRecord::Base
...
   validates :dockerfile, presence: true, unless: :external?
-  [:dockerfile, :image_name].each do |attribute|
-    validates(
-      :git_sha,
-      allow_nil: true,
-      uniqueness: {scope: attribute, message: "already exists with this #{attribute}"},
-      if: attribute
-    )
+  [:git_sha, :external_url].each do |attribute|
+    [:dockerfile, :image_name].each do |scope|
```
Comment (grosser) [inline app/models/build.rb]: that's already merged ? <-> rebase ?

**Comment by adammw** [comment]
yeah, it is already lazy-loaded, but definitely thinking of not showing it unless more than one build is detected, this is just getting the valuable bit out the door... the js is doing the changing of the visible builds when the ref changes, the validation is still in the server-side, however some of it is spread between record validation and job execution since we don't actually know the resolved sha until the deploy is running.

the other problem is that the builds displayed aren't the same as what build finder would find, need to think about how to deal with that.

again, this is because with samson we're trying to deploy the code commit not the artefact that it makes this so ugly/difficult.

File: app/helpers/builds_helper.rb (hunk 2 above)
Comment (adammw) [reply app/helpers/builds_helper.rb]: build.status always returns a string, defaults to `"not built"` - same as before

File: app/models/build.rb (hunk 3 above)
Comment (adammw) [reply app/models/build.rb]: #3510 isn't merged yet, but yes they are the same commits, will disappear once merged.

**Comment by grosser** [comment]
let's get that merged / tested / deployed before we go deeper :)

**Pull request closed without merging**